test/unittests.cpp

```diff
@@ -145,6 +145,21 @@ void TestMethodRecursion()
     Assert(ProgramOutput == correctOutput );
 }
 
+void TestComments()
```

Comment:
I think you forgot to add this to the Tests `std::vector` down at the bottom. When you do and rerun the test suite, you'll likely get memory leaks...

The general approach you have is solid, you've just overlooked one small thing which I probably should have warned you about. A `TokenList` is a `std::vector<Token*>` and these pointers point to heap allocations. Normally, the `TokenLists` are destroyed after a program is run and the memory freed. In your case, because you're storing the abridged `TokenLists` in the `CodeLine`, when these are freed, only the `Token*`s in your `TokenList tokensNoComments` will be freed. 

A solution is to delete the excess `Token`s which represent comments in the original lists. (Be careful that you don't also delete the `Tokens` you want to keep.

PS.
C++ is weird like this. Objects/standard library classes seem to make things easier but they really also hide a lot of complexity that we as programmers do have to worry about. At its heart, C++ is just C after its taken weird growth hormones. Vectors are like self resizing arrays, but we still have to worry about manual memory management.

Reply:
fixed